### PR TITLE
Fix #26

### LINKED ISSUE #26: Detect duplicate names in object literals
The object literal

    (object [foo 1] [foo 2])

will generate {foo: 1, foo:2}.  It evaluates to the object {foo: 2}.

A warning saying that the name `foo` is used twice would be a good thing.

### PR DESCRIPTION
**Pull request by octplane**
Because I'm trying to use this lib for a little project of mine, I found out this issue and wanted to give it a stab.

That's probably my first ever racket PR, so criticism is most welcome!

### CODE REVIEW

File: urlang/main.rkt [hunk 1]
```diff
@@ -1473,12 +1473,26 @@
...
+      (remove-duplicates (map syntax-e (filter (λ(item) 
+        (>
+          (length
+            (filter (λ(other) (free-identifier=? item other)) pns))
+        1))
+      pns)
+  )))
+
+
   (with-output-language (Lur Expr)
```
Comment (soegaard) [inline urlang/main.rkt:1486]: This looks like the right idea. 

I suggest using `check-duplicates` from `racket/list` with `free-identifier=?` as "same" [1].
It confuses me slightly that the return value of `duplicates?` is a list. 
Usually functions that has a named that ends in `?` return `#f` or "a true value". 
Note that `check-duplicates` returns `#f` to mean that no duplicates were found.

[1] https://docs.racket-lang.org/reference/pairs.html#%28def._%28%28lib._racket%2Flist..rkt%29._check-duplicates%29%29

File: urlang/main.rkt [hunk 2]
```diff
@@ -1473,12 +1473,26 @@
...
     (syntax-parse o
       #:literal-sets (keywords)
       [(object [pn:PropertyName e] ...)
        (let ([pn (syntax->list #'(pn ...))]
              [e  (stx-map parse-expr #'(e  ...))])
+          (define dup (duplicates? pn))
+          (when (> (length dup) 0)
+            (displayln 
+              (format "Warning: object contains shadowing variables ~a: ~a" dup (syntax->datum o))))
          `(object [,pn ,e] ...))])))
```
Comment (soegaard) [inline urlang/main.rkt:1496]: If `duplicates?` return `#f` in the case of no duplicates, you can write

```
(unless (duplicates pn)
   ...)
```

You have opted for displaying a warning - I am wondering whether it were better to make it an error?
That is, use something like this:

    (raise-syntax-error 'parse-object (~a "object contains shadowing variables " o) o) 
    
I can't remember, is duplicate names in objects an error in JavaScript - or is it ignored?

**Review by soegaard — COMMENTED** [review]
Thanks for looking into this.

File: urlang/main.rkt (hunk 1 above)
Comment (soegaard) [reply urlang/main.rkt:1486]: I must have clicked in the wrong spot. The comment above is with regards to:

```
(define (duplicates? pns)
      (remove-duplicates (map syntax-e (filter (λ(item) 
        (>
          (length
            (filter (λ(other) (free-identifier=? item other)) pns))
        1))
      pns)
  )))
  ```

File: urlang/main.rkt (hunk 2 above)
Comment (octplane) [reply urlang/main.rkt:1496]: Duplicate are ignored in JavaScript and shadowing do happen.

**Comment by soegaard** [comment]
Looks fine - except that the error message needs to go to the current error port  instead of the current output port.
I'll merge and add the `(current-error-port)` to the display expressions.

Thanks for looking into this.

**Comment by octplane** [comment]
Thanks!

**Pull request merged**